File: app/src/main/java/com/agarron/simpleast/MainActivity.java
package com.agarron.simpleast;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.SpannableStringBuilder;
import android.text.style.CharacterStyle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.agarron.simpleast_core.node.Node;
import com.agarron.simpleast_core.node.StyleNode;
import com.agarron.simpleast_core.node.TextNode;
import com.agarron.simpleast_core.parser.ParseSpec;
import com.agarron.simpleast_core.parser.Parser;
import com.agarron.simpleast_core.parser.Rule;
import com.agarron.simpleast_core.renderer.SpannableRenderableNode;
import com.agarron.simpleast_core.renderer.SpannableRenderer;
import com.agarron.simpleast_core.simple.SimpleMarkdownRules;
import com.agarron.simpleast_core.simple.SimpleRenderer;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;

public class MainActivity extends AppCompatActivity {

  private TextView resultText;
  private EditText input;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    resultText = (TextView) findViewById(R.id.result_text);
    input = (EditText) findViewById(R.id.input);

    input.setText("*t*");

    findViewById(R.id.benchmark_btn).setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        final double times = 50.0;
        long totalDuration = 0L;
        for (int i = 0; i < times; i++) {
          final long start = System.currentTimeMillis();
          testParse(50);
          final long end = System.currentTimeMillis();
          final long duration = end - start;
          totalDuration += duration;
          Log.d("timer", "duration of parse: " + duration + " ms");
        }
        Log.d("timer", "average parse time: " + totalDuration / times + " ms");
      }
    });

    findViewById(R.id.test_btn).setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        final Parser<SpannableRenderableNode> parser = new Parser<>(true);
        parser.addRules(SimpleMarkdownRules.createSimpleMarkdownRules());
        final Collection<SpannableRenderableNode> nodes = parser.parse(input.getText());
        resultText.setText(SpannableRenderer.render(new SpannableStringBuilder(), nodes, null));
      }
    });
  }

  private List<Rule<SpannableRenderableNode>> getRules() {
    final List<Rule<SpannableRenderableNode>> rules = SimpleMarkdownRules.createSimpleMarkdownRules(false);

    final Rule<SpannableRenderableNode> replacementTextRule = new Rule<SpannableRenderableNode>(SimpleMarkdownRules.INSTANCE.getPATTERN_TEXT(), true) {
      @Override
      public ParseSpec<? extends SpannableRenderableNode> parse(Matcher matcher, Parser<SpannableRenderableNode> parser, boolean isNested) {
        if (isNested) {
          return ParseSpec.createTerminal(new TextNode(matcher.group()));
        } else {
          final String target = matcher.group().replace("youtube", "YOUTUBEWASHERE");
          final List<SpannableRenderableNode> innerNodes = parser.parse(target, true);
          final StyleNode parentNode = new StyleNode(Collections.<CharacterStyle>emptyList());
          for (final Node child : innerNodes) {
            parentNode.addChild(child);
          }

          return ParseSpec.createTerminal(parentNode);
        }
      }
    };

    rules.add(replacementTextRule);

    return rules;
  }

  private String createTestText() {
    return "\u001B[0;31mERROR:\u001B[0m Signature extraction failed: Traceback (most recent call last):\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/youtube.py\", line 1011, in _decrypt_signature\n" +
        "    video_id, player_url, s\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/youtube.py\", line 925, in _extract_signature_function\n" +
        "    errnote='Download of %s failed' % player_url)\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py\", line 519, in _download_webpage\n" +
        "    res = self._download_webpage_handle(url_or_request, video_id, note, errnote, fatal, encoding=encoding, data=data, headers=headers, query=query)\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py\", line 426, in _download_webpage_handle\n" +
        "    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query)\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py\", line 406, in _request_webpage\n" +
        "    return self._downloader.urlopen(url_or_request)\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/YoutubeDL.py\", line 2000, in urlopen\n" +
        "    req = sanitized_Request(req)\n" +
        "  File \"/usr/local/lib/python3.5/dist-packages/youtube_dl/utils.py\", line 518, in sanitized_Request\n" +
        "    return compat_urllib_request.Request(sanitize_url(url), *args, **kwargs)\n" +
        "  File \"/usr/lib/python3.5/urllib/request.py\", line 269, in init\n" +
        "    self.full_url = url\n" +
        "  File \"/usr/lib/python3.5/urllib/request.py\", line 295, in full_url\n" +
        "    self._parse()\n" +
        "  File \"/usr/lib/python3.5/urllib/request.py\", line 324, in _parse\n" +
        "    raise ValueError(\"unknown url type: %r\" % self.full_url)\n" +
        "ValueError: unknown url type: '/yts/jsbin/player-en_US-vflkk7pUE/base.js'\n" +
        " (caused by ValueError(\"unknown url type: '/yts/jsbin/player-en_US-vflkk7pUE/base.js'\",))";
  }

  private void testParse(final int times) {
    final String text = createTestText();

    for (int i = 0; i < times; i++) {
      SimpleRenderer.renderBasicMarkdown(resultText, text);
    }
  }
}

File: app/src/main/java/com/discord/simpleast/sample/MainActivity.kt
package com.discord.simpleast.sample

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.text.SpannableStringBuilder
import android.util.Log
import android.widget.EditText
import android.widget.TextView
import com.agarron.simpleast.R
import com.discord.simpleast.core.parser.ParseSpec
import com.discord.simpleast.core.parser.Parser
import com.discord.simpleast.core.parser.Rule
import com.discord.simpleast.core.renderer.SpannableRenderableNode
import com.discord.simpleast.core.renderer.SpannableRenderer
import com.discord.simpleast.core.simple.SimpleMarkdownRules
import com.discord.simpleast.core.simple.SimpleRenderer
import java.util.regex.Matcher
import java.util.regex.Pattern

class MainActivity : AppCompatActivity() {

  private lateinit var resultText: TextView
  private lateinit var input: EditText

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    resultText = findViewById(R.id.result_text) as TextView
    input = findViewById(R.id.input) as EditText

    input.setText("*t*")

    findViewById(R.id.benchmark_btn).setOnClickListener {
      val times = 50.0
      var totalDuration = 0L
      var i = 0
      while (i < times) {
        val start = System.currentTimeMillis()
        testParse(50)
        val end = System.currentTimeMillis()
        val duration = end - start
        totalDuration += duration
        Log.d("timer", "duration of parse: $duration ms")
        i++
      }
      Log.d("timer", "average parse time: " + totalDuration / times + " ms")
    }

    findViewById(R.id.test_btn).setOnClickListener {
      testParseWithContext(input.text, resultText)
    }
  }

  private fun testParseWithContext(source: CharSequence?, resultText: TextView) {
    val parser = Parser<SpannableRenderableNode<TestRenderContext>>()

    parser.addRule(object : Rule<SpannableRenderableNode<TestRenderContext>>(Pattern.compile("^<replacement>")) {
      override fun parse(matcher: Matcher, parser: Parser<SpannableRenderableNode<TestRenderContext>>, isNested: Boolean): ParseSpec<out SpannableRenderableNode<TestRenderContext>> {
        return ParseSpec.createTerminal(TextNodeWithContext())
      }
    })

    parser.addRules(SimpleMarkdownRules.createSimpleMarkdownRules())

    val ast = parser.parse(source)
    resultText.text = SpannableRenderer.render(SpannableStringBuilder(), ast, TestRenderContext("poooooooop"))
  }

  data class TestRenderContext(val string: String)

  private fun createTestText() = """[0;31mERROR:[0m Signature extraction failed: Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/youtube.py", line 1011, in _decrypt_signature
    video_id, player_url, s
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/youtube.py", line 925, in _extract_signature_function
    errnote='Download of %s failed' % player_url)
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py", line 519, in _download_webpage
    res = self._download_webpage_handle(url_or_request, video_id, note, errnote, fatal, encoding=encoding, data=data, headers=headers, query=query)
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py", line 426, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query)
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/extractor/common.py", line 406, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/YoutubeDL.py", line 2000, in urlopen
    req = sanitized_Request(req)
  File "/usr/local/lib/python3.5/dist-packages/youtube_dl/utils.py", line 518, in sanitized_Request
    return compat_urllib_request.Request(sanitize_url(url), *args, **kwargs)
  File "/usr/lib/python3.5/urllib/request.py", line 269, in init
    self.full_url = url
  File "/usr/lib/python3.5/urllib/request.py", line 295, in full_url
    self._parse()
  File "/usr/lib/python3.5/urllib/request.py", line 324, in _parse
    raise ValueError("unknown url type: %r" % self.full_url)
ValueError: unknown url type: '/yts/jsbin/player-en_US-vflkk7pUE/base.js'
 (caused by ValueError("unknown url type: '/yts/jsbin/player-en_US-vflkk7pUE/base.js'",))"""

  private fun testParse(times: Int) {
    val text = createTestText()

    for (i in 0 until times) {
      SimpleRenderer.renderBasicMarkdown(text, resultText)
    }
  }

  private class TextNodeWithContext : SpannableRenderableNode<TestRenderContext>() {
    override fun render(builder: SpannableStringBuilder, renderContext: TestRenderContext) {
      builder.append(renderContext?.string)
    }
  }
}

